src/binary/timestamp.rs

```diff
@@ -0,0 +1,98 @@
+// Copyright Amazon.com, Inc. or its affiliates.
+
+use std::io::{self, Write};
+
+use bytes::BufMut;
+use chrono::{Datelike, Timelike};
+use smallvec::SmallVec;
+
+use crate::{
+    binary::{uint::DecodedUInt, var_int::VarInt, var_uint::VarUInt, writer::MAX_INLINE_LENGTH},
+    result::IonResult,
+    types::timestamp::{Precision, Timestamp},
+};
+
+/// Extends [`Timestamp`] to support [Ion binary].
+///
+/// [Ion binary]: https://amzn.github.io/ion-docs/docs/binary.html#6-timestamp
+impl Timestamp {
+    const BUFFER_SIZE: usize = 16;
+
+    /// NOTE: Currently, this function always encodes with nanosecond precision.
+    pub fn encode_binary(&self) -> IonResult<SmallVec<[u8; Timestamp::BUFFER_SIZE]>> {
+        const SECONDS_PER_MINUTE: f32 = 60f32;
+        let mut buffer = [0u8; Timestamp::BUFFER_SIZE];
+        let mut writer = io::Cursor::new(&mut buffer).writer();
+
+        // Each component of the timestamp is in UTC time. Readers then apply the offset minutes
+        // to derive the localized time.
+        let utc = self.date_time;
+
+        // Ion encodes offsets in minutes while DateTime stores it in seconds.
+        let offset_seconds = self.offset.map(|o| o.utc_minus_local()).unwrap_or(0);
+        let offset_minutes = (offset_seconds as f32 / SECONDS_PER_MINUTE).round() as i64;
+        VarInt::write_i64(&mut writer, offset_minutes)?;
+
+        VarUInt::write_u64(&mut writer, utc.year() as u64)?;
+        VarUInt::write_u64(&mut writer, utc.month() as u64)?;
+
+        // So far, we've written required fields. The rest are optional!
```

Comment:
Month [is an optional field](https://amzn.github.io/ion-docs/docs/binary.html#6-timestamp) as well:

> Timestamp representations have 7 components, where 5 of these components are optional depending on the precision of the timestamp. The 2 non-optional components are offset and year. The 5 optional components are (from least precise to most precise): month, day, hour and minute, second, fraction_exponent and fraction_coefficient. All of these 7 components are in Universal Coordinated Time (UTC).

Reply:
Whoops. Thanks.